src/json.ts

```diff
@@ -34,6 +41,119 @@ export type Json =
   | Json[]
   | { [prop: string]: Json };
 
+/**
+ * A helper type to make properties with `undefined` in their type optional, but
+ * not `undefined` itself.
+ *
+ * @example
+ * ```ts
+ * type Foo = ObjectOptional<{ foo: string | undefined }>;
+ * // Foo is equivalent to { foo?: string }
+ * ```
+ */
+export type ObjectOptional<Schema extends Record<string, unknown>> = {
+  [Key in keyof Schema as Schema[Key] extends ExactOptionalGuard
+    ? Key
+    : never]?: Schema[Key] extends ExactOptionalGuard & infer Original
+    ? Original
+    : never;
+} & {
+  [Key in keyof Schema as Schema[Key] extends ExactOptionalGuard
+    ? never
+    : Key]: Schema[Key];
+};
+
+/**
+ * An object type with support for exact optionals. This is used by the `object`
+ * struct. This uses the {@link ObjectOptional} helper to make properties with
+ * `undefined` in their type optional, but not `undefined` itself.
+ */
+export type ObjectType<Schema extends ObjectSchema> = Simplify<
+  ObjectOptional<
+    Optionalize<{
+      [Key in keyof Schema]: Infer<Schema[Key]>;
+    }>
+  >
+>;
+
+/**
+ * A struct to check if the given value is a valid object, with support for
+ * {@link exactOptional} types.
+ *
+ * @param schema - The schema of the object.
+ * @returns A struct to check if the given value is an object.
+ */
+export const object = <Schema extends ObjectSchema>(
+  schema: Schema,
+): Struct<ObjectType<Schema>> =>
+  // The type is slightly different from a regular object struct, because we
+  // want to make properties with `undefined` in their type optional, but not
+  // `undefined` itself. This means that we need a type cast.
+  superstructObject(schema) as unknown as Struct<ObjectType<Schema>>;
+
+declare const exactOptionalSymbol: unique symbol;
+type ExactOptionalGuard = {
+  _exactOptionalGuard?: typeof exactOptionalSymbol;
```

Comment:
Can you explain how this type is used? I see that it's referred to above, but where does  this property get set?

Reply:
This property is never set: It's only used for inferring types. It's added to the type of the exact optional struct, and used by `ObjectOptional` to infer the proper type.

Reply:
Okay. This is advanced! I can see how that works though.

Reply:
Yeah it's definitely advanced. We use a `unique symbol` type to ensure that the type cannot be replicated in any way. `unique symbol`s are guaranteed to be unique:

```ts
declare const a: unique symbol;
declare const b: unique symbol;

// TS2367: This condition will always return  false  since the types  typeof a  and  typeof b  have
// no overlap.
a === b;
```